### PR TITLE
Implement RESP3 set type for the set data structure

### LINKED ISSUE #1997: Add RESP3 set type
Please refer [Redis RESP3](https://github.com/redis/redis-specifications/blob/master/protocol/RESP3.md)

### PR DESCRIPTION
**Pull request by git-hulk**
This closes #1997

### CODE REVIEW

**Comment by caipengbo** [comment]
I have a question about  last PR #2017, why put `MultiBulkString` to `Connection`? @git-hulk

**Comment by git-hulk** [comment]
> I have a question about last PR #2017, why put `MultiBulkString` to `Connection`? @git-hulk

@caipengbo Because `MultiBulkString` should use the RESP3's null to represent the nil string if RESP3 was enabled. That said, all types related to RESP3 will move into `Connnection`.

**Review by torwig — DISMISSED** [review]
LGTM.

BTW, the line:
` *output = conn->SetLen(3);`
made me stop and think for a while :) ("How do we set a length to a connection?")

**Comment by git-hulk** [comment]
> LGTM.
> 
> BTW, the line: ` *output = conn->SetLen(3);` made me stop and think for a while :) ("How do we set a length to a connection?")
 
Yes, it's a bit confusing. Do you have any idea about this name? I'm not sure if it's better to append `Build` before those functions like `BuildSetLen()`, `BuildMultiBulkStrings()`

@torwig What do you think about this solution?

**Comment by caipengbo** [comment]
> MultiString should use the RESP3's null to represent the nil string if RESP3 was enabled.

Why not pass a version flag to this function?

**Comment by git-hulk** [comment]
> > MultiString should use the RESP3's null to represent the nil string if RESP3 was enabled.
> 
> Why not pass a version flag to this function?

It's okay to add the version flag, but moving them into connection can avoid misusing those functions. That said, use RESP2 or RESP3 should be related to connection instead of developers.

**Comment by caipengbo** [comment]
> That said, use RESP2 or RESP3 should be related to connection instead of developers.

Got it !

**Comment by torwig** [comment]
@git-hulk The `Build` prefix doesn't fit because the function doesn't build anything. I'm thinking of something like `SizeOfSet` similar to the `ArrayOfSet` function you introduced as well.

**Comment by git-hulk** [comment]
> @git-hulk The `Build` prefix doesn't fit because the function doesn't build anything. I'm thinking of something like `SizeOfSet` similar to the `ArrayOfSet` function you introduced as well.

cool. SizeOfSet is better, I have updated it.

**Review by caipengbo — APPROVED** [review]
LGTM

**Review by torwig — APPROVED** [review]

File: src/server/redis_connection.cc [hunk 1]
```diff
@@ -163,6 +163,15 @@ std::string Connection::MultiBulkString(const std::vector<std::string> &values,
   return result;
 }
 
+std::string Connection::ArrayOfSet(const std::vector<std::string> &elems) const {
```
Comment (PragmaTwice) [inline src/server/redis_connection.cc:166]: IMHO I cannot understand the name.

It is just a **set**, not an **array of sets**. Instead, it is a **set of bulk strings**.

An **array of sets** is typed `Array<Set<T>>`.

Naming is important, and these confusing names can lead to bugs.

Comment (git-hulk) [reply src/server/redis_connection.cc:166]: @PragmaTwice So do you suggest changing this name to SetOfBulkStrings? If yes, I will bring this fix in the next RESP3 PR.

Comment (PragmaTwice) [reply src/server/redis_connection.cc:166]: Yeah that's better to me. ArrayOfSet seems like an array composed by sets.

Comment (git-hulk) [reply src/server/redis_connection.cc:166]: Thanks for your good suggestion.

**Pull request merged**